[1단계 - 자동차 경주 구현] 배카라(박성우) 미션 제출합니다.

seong-wooo opened the pull request:
# 안녕하세요. 닉! 귀한 시간 내주셔서 감사합니다.. ! 
작년에 개발바닥 유튜브의 이력서 영상에서 봤었는데 코드리뷰를 받게 되어 영광입니다 !

### 미션을 진행하면서 궁금했던 점들이 있었습니다.

1. 테스트 코드를 작성하면서 테스트를 진행하기 위해서 private를 public으로 변경하였는데요.
테스트를 위해서 접근 제한자를 변경하는게 옳은 지, 아니면 그냥 테스트를 진행하지 않고 private로 작성하는 게 좋은지 잘 모르겠습니다.
어떤 것이 더 옳을까요??

2. 우승자_찾기() 라는 테스트 메서드를 만들기 위해 Car 클래스에 새로운 생성자를 생성하였습니다.
테스트만을 위해서 새로운 코드를 만드는 것은 괜찮을까요??

3. 왜인지 모르게 getter를 사용하는 것은 자존심이 허락하지 않는 것 같습니다.
그래서 getter를 사용하지 않고 객체에서 정보를 얻어오기 위해 많은 생각을 하였는데요.
예를 들어, 우승한 자동차들의 이름을 알아내는 부분에서 
getName()을 통해 이름을 가져오는 방식이 아닌,
appendName()이라는 메서드로 리스트에 이름을 담아오는 방식을 사용했습니다.
getter의 사용을 피하려고 생각한 방법이지만 약간 찝찝한 느낌이 있습니다. 이 방법이 getter와 큰 차이가 있나 싶습니다.
우승자들의 이름을 알기 위해 어떤 방법을 사용했어야 좋을까요??

4. Service 패키지 내에 RandomNumberService라는 클래스를 만들어서 랜덤 값을 반환하는 메서드를 작성하였습니다.
도메인에 관련된 코드가 아니더라도 Service라는 이름을 붙여서 따로 클래스를 만들어도 괜찮을까요??

쓰다보니 질문이 조금 길어졌습니다.. ㅎ ㅎ 
저를 리뷰로 혼내주세요.
감사합니다.

seong-wooo committed:
docs: README 기능 구현 목록 작성

seong-wooo committed:
feat: 입력받은 문자열 null 또는  일 경우 0을 반환

seong-wooo committed:
feat: 숫자 하나만 입력 받았을 경우 그 숫자를 반환

seong-wooo committed:
feat: 컴마(,)구분자로 분리한 뒤 합을 반환

seong-wooo committed:
feat: 구분자에 콜론 추가

seong-wooo committed:
feat: 커스텀 구분자 기능 추가

seong-wooo committed:
feat: 음수 입력 시 RuntimeException 발생

seong-wooo committed:
refactor: 변수명 변경

seong-wooo committed:
style: 컨벤션에 맞게 공백추가

seong-wooo committed:
docs: README 구현 기능 목록 추가

seong-wooo committed:
feat: 이름으로 공백을 입력받은 경우 예외처리

seong-wooo committed:
feat: 경주할 자동차들 이름을 입력받는 기능 구현

seong-wooo committed:
feat: 자동차 이름 구분하는 기능 구현

seong-wooo committed:
feat: 시도할 횟수 입력받는 기능 구현

seong-wooo committed:
feat: 자동차 이름이 5자 초과일 경우 예외처리

seong-wooo committed:
feat: 시도 횟수가 숫자가 아닌 경우 예외처리

seong-wooo committed:
feat: 시도 횟수가 양수가 아닌 경우 예외처리

seong-wooo committed:
feat: 이름으로 자동차 객체 생성

seong-wooo committed:
feat: 자동차 리스트 객체 구현

seong-wooo committed:
feat: 게임턴 객체 구현 및 OutputView 구현

seong-wooo committed:
feat: 전진 조건 확인 기능 구현

seong-wooo committed:
refator: 전진 조건 확인 기능 수정

seong-wooo committed:
feat: 조건에 따른 전진 기능 구현

seong-wooo committed:
feat: 남은 턴이 있는지 확인하는 기능 구현

seong-wooo committed:
docs: 기능 구현 목록 추가

seong-wooo committed:
feat: 횟수 차감하는 기능 구현

seong-wooo committed:
feat: 턴마다 자동차 이동시키는 기능 구현

seong-wooo committed:
feat: 턴마다 자동차들의 위치 상태 출력 기능 구현

seong-wooo committed:
feat: 우승자 명단 찾기 기능 구현

seong-wooo committed:
feat: main 함수 구현

seong-wooo committed:
refactor: 메서드 순서 변경, 상수화, 메서드명 변경

seong-wooo committed:
refactor: 필드 변수 접근 제한자 변경, final 키워드 사용

wbluke commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,60 @@
+
+	public boolean checkMovingCondition(int randomNumber) {
+		return randomNumber >= MOVING_CONDITION_NUMBER;
+	}
```
> 테스트 코드를 작성하면서 테스트를 진행하기 위해서 private를 public으로 변경하였는데요.
테스트를 위해서 접근 제한자를 변경하는게 옳은 지, 아니면 그냥 테스트를 진행하지 않고 private로 작성하는 게 좋은지 잘 모르겠습니다.
어떤 것이 더 옳을까요??

private 메서드에 대한 테스트가 필요한지에 대한 질문으로 바꿔서 생각해도 될 것 같은데요 ㅎㅎ
아래 사이트로 답변 드릴게요 :)
http://shoulditestprivatemethods.com/
(개발자도구를 열어서 소스를 보시면 추가 설명을 보실 수 있습니다ㅎㅎ)

private 메서드를 테스트하고 싶어진다면 다음을 고려해보아야 합니다!
- 상위 public 메서드에서 테스트가 충분하게 이루어지지 않았는가? (해당 private 메서드를 포함한 테스트를 진행하지 않았는가?)
- 하나의 객체가 너무 많은 일을 하고 있지는 않은가? 객체의 책임을 분리할 시점이 아닌가?

wbluke commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	public void move() {
+		if (checkMovingCondition(RandomNumberService.getRandomNumber())) {
```
이 public 메서드의 테스트를 위해서 랜덤값 생성 로직을 외부로 분리해보면 어떨까요?
테스트가 어려운 부분을 외부로 분리하면 코드도 명확해지고, 테스트도 가능해질 것 같아요 :)

wbluke commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,60 @@
+	public Car(String name, int position) {
+		this.name = name;
+		this.position = position;
+	}
```
> 우승자_찾기() 라는 테스트 메서드를 만들기 위해 Car 클래스에 새로운 생성자를 생성하였습니다.
테스트만을 위해서 새로운 코드를 만드는 것은 괜찮을까요??

기본적으로 프로덕션 코드가 아니라 테스트만을 위해서 새로운 코드를 추가하는 것은 지양하는 것이 맞습니다.
다만 지금 같은 경우 테스트를 위해서 `생성자를 추가`하는 경우라면 다른 방향으로 고민해볼 수도 있는데요 ㅎㅎ
해당 생성자가 미래 시점에 프로덕션 코드에서 쓰일 수도 있다고 판단한다면 추가할 수도 있을 것이고,
다음과 같이 생성자 간 활용을 하여 자연스러운 코드를 만들어볼 수도 있겠네요 :)
```java
public Car(String name) {
    this(name, 0);
}
```

wbluke commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -0,0 +1,60 @@
+
+	public void appendName(List<String> winnerNames) {
+		winnerNames.add(name);
+	}
```
> 왜인지 모르게 getter를 사용하는 것은 자존심이 허락하지 않는 것 같습니다.
그래서 getter를 사용하지 않고 객체에서 정보를 얻어오기 위해 많은 생각을 하였는데요.
예를 들어, 우승한 자동차들의 이름을 알아내는 부분에서
getName()을 통해 이름을 가져오는 방식이 아닌,
appendName()이라는 메서드로 리스트에 이름을 담아오는 방식을 사용했습니다.
getter의 사용을 피하려고 생각한 방법이지만 약간 찝찝한 느낌이 있습니다. 이 방법이 getter와 큰 차이가 있나 싶습니다.
우승자들의 이름을 알기 위해 어떤 방법을 사용했어야 좋을까요??

오 getter를 사용할 때 자존심이 허락하지 않는 이유가 궁금하네요 ㅎㅎ
getter를 사용하지 않으려는 이유가 객체지향 생활 체조의 `9. getter/setter/property를 쓰지 않는다.` 때문이라면, getter를 쓰지 않는다는 말의 의미를 다시 생각해볼 필요가 있습니다.

프로그램은 `데이터`와 `코드`의 조합입니다.
데이터가 필요한 시점에서는 getter를 사용하는 것이 가장 자연스럽습니다.
그 외의 경우, 즉 getter를 사용하지 않고 객체 간에 메시지를 보내 소통함으로 해결할 수 있는 것들은 그렇게 설계하는 것이 가장 좋습니다 :)

wbluke commented on src/main/java/racingcar/service/RandomNumberService.java:
```diff
@@ -0,0 +1,8 @@
+package racingcar.service;
+
+public class RandomNumberService {
+	public static int getRandomNumber() {
+		int random = (int)(Math.random() * 10);
+		return random;
+	}
```
> Service 패키지 내에 RandomNumberService라는 클래스를 만들어서 랜덤 값을 반환하는 메서드를 작성하였습니다.
도메인에 관련된 코드가 아니더라도 Service라는 이름을 붙여서 따로 클래스를 만들어도 괜찮을까요??

네이밍에 정답이 있는 것은 아니죠 ㅎㅎ
레이어드 아키텍처를 의식하여 질문을 주신거라면 크게 상관하지 않으셔도 될 것 같습니다 :)

wbluke commented on src/main/java/racingcar/controller/RacingCarGame.java:
```diff
@@ -0,0 +1,101 @@
+		List<String> carNames = splitCarNames(inputCarNames);
+		try {
+			checkCarNamesBlank(inputCarNames);
+			checkCarNamesLength(carNames);
```
유효성 검사를 메인 클래스에서 하다보니 로직이 많이 길어지고 필요한 유효성 검사가 늘어난다면 유지보수도 어려워질 것 같아요 ㅎㅎ
그리고 해당 유효성에 대한 테스트도 크기가 큰 RacingCarGame 객체를 테스트해야만 체크가 가능합니다.

Car에서 객체 생성 시점에 유효성을 검사해보면 어떨까요?
```java
Car car = new Car("123456"); // thow exception
```
이렇게 되면 `생성자`라는 것의 의미도 확실해지고, Car 단위테스트도 가능할 것 같아요 :)

다른 유효성체크도 한번 고민해보세요! :)
(Cars 일급컬렉션을 만들어주셔서 말인데, 다수의 자동차를 생성하고 검증하는 역할도 Cars에서 해볼 수 있지 않을까요?)

seong-wooo replied:
gameTurn 또한 컨트롤러에서 입력을 받고 GameTurn 클래스 내부에서 검증을 진행하고자 합니다. 
검증을 GameTurn 내부에서 진행하는 것이 gameTurn 객체의 역할인 것 같습니다.
gameTurn은 숫자인지 확인을 해봐야합니다.
그런데 입력을 받은 값을 gameTurn 생성자로 그대로 넘기게 되면 

```
public GameTurn(String gameTurn) {
	checkGameTurnNumber(gameTurn);
	this.gameTurn = Integer.parseInt(gameTurn);
}
```

위와 같이 작성을 하게 됩니다.
this.gameTurn의 자료형은 int형 이기때문에
생성자의 인자로 String 자료형이 들어가는 것에 대해서 어색함을 느낍니다.
이렇게 된다면.. 그냥 원래처럼 컨트롤러에서 gameTurn 입력에 대한 검증을 하는 것이 맞을까요?
아니면 gameTurn 내부에서 검증을 하는 것이 맞을까요?

seong-wooo replied:
Cars 클래스에서 Car 객체를 생성하도록 변경하려고 합니다.
그런데 Cars 클래스에서 Car 객체 생성을 담당하도록 할 때 
컨트롤러의 역할이 어디까지인지 잘 모르겠습니다. 
Car 객체를 생성하기까지의 과정이 

1.사용자에게 이름들을 입력받는다. (String) -> 2. 이름들을 "," 문자 기준으로 자른다. (List<String>)-> 3. 자른 이름을 이용하여 Car 객체를 만든다. (List<Car>)

1,2,3으로 나눌 수 있을 것 같습니다.

Cars에서 Car 객체를 만들기 위해 
컨트롤러에서 1번을 수행 후 Cars 객체에게 입력받은 이름을 건네줄 지,
아니면 컨트롤러에서 2번(자르기)를 실행하고 건네줄 지 모르겠습니다.

제 생각에 이 문제를 고민하는 이유는 위의 질문과 동일합니다.

이 또한 생성자에게 Cars 클래스의 필드변수의 자료형(List<Car>)과는 다른 자료형을 넘겨야한다는 어색함이 생깁니다.

예를 들어

```
public Cars(String carNames) {
	this.cars = makeCars(carNames); 
}

또는 

public Cars(List<String> carNames) {
	this.cars = makeCars(carNames); 
}
```
계속된 같은 질문을 하는 것으로 보아 제가 놓치고 있는 무엇인가가 있는 것 같습니다.
어색하지 않은데 혼자서 어색함을 느끼는 것인지,, 
그냥 잘못된 코드를 작성하고 있는 것인지,,

저를 혼내주세요..

wbluke replied:
아주 좋은 질문입니다!
이제 여기서 유효성검증의 성격을 구분해보면 좋을 것 같아요.

#### 게임 횟수 validation

1. 타입이 int인지 체크한다.
2. 1 이상인지 체크한다.

#### 자동차 리스트 validation

1. 구분자(,)로 구분되는 문자열인지 체크하고 나눈다.
2. 각 자동차 이름은 최대 5자까지 가능하다.

---
각 validation 중 `도메인(모델) 로직`이라고 볼 수 있는 validation은 무엇일까요?
배카라가 어색함을 느꼈던 지점처럼, 각 케이스의 1번 validation은 어떻게보면 Model의 영역이 아니라 View나 Controller의 역할이라고 생각해요.
(저는 엄밀히 말해서 View의 역할이라고 생각해요.)
View 쪽에서 타입이나 사용자 입력에 대한 기본적인 유효성은 보장을 해준 다음, 
숫자가 **횟수**라는 의미를 갖기 위해서 1 이상인지, 문자열이 정해진 자동차 이름의 규칙에 맞기 위해서 5자 이상인지를 확인하는 것은 각 도메인 객체의 생성자에서 보장해주어야 할 것 같아요.
결국 횟수라는 숫자는 1 이상이어야 한다는 규칙이나, 자동차 이름이 5자 이하라는 규칙은 우리가 만드는 콘솔 게임의 핵심 도메인 로직이라고 볼 수 있기 때문이에요.

---
조금 다른 관점에서 볼까요?
GameTurn과 Cars라는 객체를 사용하는 클라이언트 입장에서 생각해봅시다.
말씀해주신 것처럼, GameTurn 생성자에서 String을 받고 있다면 해당 객체를 생성해서 사용하려는 클라이언트가 어떤 값을 넣어야할지 명확하게 알 수 있을까요?
Cars 생성자에서 단순한 String을 받고 있다면 Cars를 생성해서 사용하려는 클라이언트가 어떤 문자열을 넣어야할지 알 수 있을까요?
이런 관점에서 생각해본다면 오히려 답이 더 명확해질 것 같아요 :)

wbluke commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,42 @@
+			throw new RuntimeException();
+		}
+
+		return numbers.stream().mapToInt(number -> number).sum();
```
- stream은 한줄에 하나씩 선언하는 것이 좋습니다.
- if문에 긴 로직이 들어갈 경우, 메서드 분리를 하여 추상화를 고려해보세요 :)

wbluke commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	public static List<String> splitText(String text) {
+		Matcher m = Pattern.compile("//(.)\n(.*)").matcher(text);
```
Pattern은 생성 비용이 큰 객체입니다.
재사용을 고려해보아요! :)

https://a1010100z.tistory.com/167

wbluke reviewed (changes requested):
안녕하세요 배카라! 자동차 리뷰를 맡은 닉입니다. 잘 부탁드려요 ㅎㅎ
여러가지를 고민하면서 잘 구현해주신 것 같네요! 👍 
몇 가지 코멘트 남겼으니 확인해주시고, 리팩토링과 테스트 추가 해보면 좋을 것 같아요 🙂

seong-wooo committed:
refactor: Car name을 검증하는 로직을 Car 내부에 구현

seong-wooo committed:
refactor: gameTurn 입력을 검증하는 로직을 gameTurn 클래스 내부에 구현

seong-wooo committed:
refactor: Car 클래스의 move 메서드 수정

move 메서드 테스트를 위해 수정

seong-wooo committed:
refactor: Car객체를 Cars 클래스 내부에서 생성하도록 수정

seong-wooo committed:
refactor: findWinnerCars 메서드 수정

Car 클래스의 getName 메서드를 이용하여 우승자의 이름을 알아내도록 수정

seong-wooo committed:
refactor: 역할에 따른 test 클래스 분리

seong-wooo committed:
refactor: RANDOM_NUMBER_REGEX 변수명 변경

RANDOM_NUMBER_REGEX를 변수명으로 설정했을 떄 gameTurn과 무슨 관계인지 파악할 수 없음
따라서 GAME_TURN_NUMBER_REGEX로 변경

seong-wooo committed:
refactor: equals와 hashCode를 같이 재정의 하도록 변경

seong-wooo committed:
refactor: Pattern 객체를 초기에 캐싱하도록 변경

seong-wooo commented:
첫 리뷰를 받고 궁금한 질문 사항을 [댓글](https://github.com/woowacourse/java-racingcar/pull/272#discussion_r805125612)로 남겼습니다. 감사합니다 :)

wbluke commented on src/main/java/racingcar/controller/RacingCarGame.java:
```diff
@@ -1,93 +1,38 @@
-	public List<String> getCarNames() {
-		String inputCarNames = inputView.getCarNames();
-		List<String> carNames = splitCarNames(inputCarNames);
+		cars = new Cars(null);
```
현재 Cars를 필드에서 상태 관리하고 있기 때문에 이런 로직이 나온 것 같아요.
null을 넣는게 (어색하긴 하지만) 문제라기 보다는,
RacingCarGame이 Cars의 상태를 가지게 된다면, 나중에 여러 사용자가 해당 게임을 동시에 진행하거나 하는 경우 의도한대로 동작하지 않을 것 같아요.
게임을 진행할 때마다 Cars를 생성하고 필드에 캐싱할 필요는 없어 보이네요 :)

(그렇게 보면 GameTurn도 마찬가지일 것 같아요 ㅎㅎ)

wbluke commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -1,60 +1,82 @@
 
-	public void move() {
-		if (checkMovingCondition(RandomNumberService.getRandomNumber())) {
+	public void move(int number) {
+		if (number >= MOVING_CONDITION_NUMBER) {
```
제가 말한 부분을 잘 캐치하고 테스트도 잘 작성해 주셨네요 ㅎㅎ
(테스트는 전진 성공 테스트와 전진 실패 테스트 모두 작성해주면 좋을 것 같아요.)

여기서 조금 더 나아가보자면, `어떤 숫자가 4보다 크다` 라는 **자동차 전진 조건**을 추상화해서 boolean을 반환하는 형태의 메서드를 가지는 인터페이스로 만들고,
파라미터로 해당 인터페이스 구현체(람다)를 받아볼 수도 있어요.
테스트에서는 테스트하고자 하는 조건을 인터페이스 구현체로 넣어줄 수 있고요.

현재 시점에서 너무 어렵게 다가올수도 있는데, 반드시 하실 필요는 없고 한번 가볍게 고민해보셔도 좋을 것 같아요 :)

wbluke reviewed (approved):
피드백 반영 잘 해주셨네요 👍 
질문주신 부분 포함해서 코멘트 남겼으니 다음 단계 진행하시면서 반영 부탁드려요 🙂
